### DISCUSSION TITLE
Having trouble deploying the Kubernetes Dashboard Helm chart

### DISCUSSION THREAD

**Discussion by nikitawootten** [hall/kubenix#27] [Q&A] [answered]
Hi, this is quite an exciting project!

I'm not sure if this is the appropriate place to post this, but I ran into an issue when creating my first kubenix configuration with the [kubernetes dashboard](https://artifacthub.io/packages/helm/k8s-dashboard/kubernetes-dashboard) helm chart:

```nix
{ kubenix ? 
  let
    lock = (builtins.fromJSON (builtins.readFile ../flake.lock)).nodes.kubenix.locked;
  in
  import (fetchTarball {
    url = "https://github.com/hall/kubenix/archive/${lock.rev}.tar.gz";
    sha256 = lock.narHash;
  }),
  system ? builtins.currentSystem
}:
(kubenix.evalModules.${system} {
  module = {kubenix, pkgs, ... }: {
    imports = with kubenix.modules; [
      k8s
      helm
    ];

    kubernetes.helm.releases.k8s-dashboard = {
      chart = kubenix.lib.helm.fetch {
        repo = "https://kubernetes.github.io/dashboard/";
        chart = "kubernetes-dashboard";
        version = "6.0.8";
        sha256 = "myLGEuL9X4yYyEbHXUMNdk+BkQ0zt0h1ybRrET5U9Xg=";
      };
    };
  };
})
```

When I try to build the output I get the following error:
```
 ➜ nix eval -f . --json config.kubernetes.generated | jq
error:
       … while calling the 'head' builtin

         at /nix/store/rbiwpy57k5wag6mvac1s0bh482q79wm1-source/lib/attrsets.nix:784:11:

          783|         || pred here (elemAt values 1) (head values) then
          784|           head values
             |           ^
          785|         else

       … while calling the 'hashString' builtin

         at /nix/store/gwpbg749y5byq40b8p202avw4hd69y4y-source/lib/k8s/default.nix:64:12:

           63|   }: let
           64|     hash = builtins.hashString "sha1" (builtins.toJSON items);
             |            ^
           65|

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: A definition for option `kubernetes.api.resources.apps.v1.Deployment.k8s-dashboard-kubernetes-dashboard.spec.template.spec.containers.kubernetes-dashboard.resources.limits.cpu' is not of type `string'. Definition values:
       - In `/nix/store/gwpbg749y5byq40b8p202avw4hd69y4y-source/modules/helm.nix': 2
```

**Comment by adrian-gierakowski** [comment, marked as answer]
looks like the default value provided by the helm char for that prop is incorrect 
try looking at the charts documentation to find out if you if it supports overriding default ` resources` and set it to correct values ("1" for 1 cpu, or an equivalent of "1000m" etc).

**Comment by hall** [comment]
Discussions are as good a place as any! Thanks for bringing this up, @nikitawootten.

As Adrian mentioned, this is because [the chart is using an int here](https://github.com/kubernetes/dashboard/blob/42deb6b32a27296ac47d1f9839a68fab6053e5fc/aio/deploy/helm-chart/kubernetes-dashboard/values.yaml#L133) so this can be worked around by overriding the value with the corresponding string:

```nix
kubernetes.helm.releases.k8s-dashboard = {
  chart = kubenix.lib.helm.fetch {
    repo = "https://kubernetes.github.io/dashboard/";
    chart = "kubernetes-dashboard";
    version = "6.0.8";
    sha256 = "myLGEuL9X4yYyEbHXUMNdk+BkQ0zt0h1ybRrET5U9Xg=";
  };
  values.resources.limits.cpu = "2";
};
```

This chart has [undergone quite a few changes](https://github.com/kubernetes/dashboard/pull/7047), which includes [using string values in the default branch](https://github.com/kubernetes/dashboard/blob/6cb01fed3b2271d3109120edc81913a6a99abe27/charts/helm-chart/kubernetes-dashboard/values.yaml#L177). So this shouldn't be an issue with the next release _but_ this still feels like a bug in kubenix as [integers are valid resource values](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-cpu).

**Reply by hall** [reply]
OK -- a little more context here. So far as I can tell.

Resource requests/limits get [set to the string type](https://github.com/hall/kubenix/blob/60a5a28db7ab569ccfd342caadd44824b64c7c93/modules/generated/v1.27.nix#L9333) during module generation because, although [the API defines the values as an `object`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#resourcerequirements-v1-core), [the swagger definition](https://raw.githubusercontent.com/kubernetes/kubernetes/v1.27.0/api/openapi-spec/swagger.json) says that `object` is specifically a `Quantity`:

```json
        "limits": {
          "additionalProperties": {
            "$ref": "#/definitions/io.k8s.apimachinery.pkg.api.resource.Quantity"
          },
          "description": "Limits describes the maximum amount of compute resources allowed. More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/",
          "type": "object"
        },
```

which is defined as

```json
    "io.k8s.apimachinery.pkg.api.resource.Quantity": {
      "type": "string"
    },
```

So [kubectl does the conversion to string itself](https://github.com/kubernetes/kubectl/blob/13f7b15bc1b75003bb91ef2d2de63c84d4525e33/pkg/util/resource/resource.go#L128-L133) which explains how, despite the API types, non-string values are supported.

With that, seems like the proper fix would to either
- cast the values to a string
- allow values to be a string _or_ integer
